main.py

```diff
@@ -1974,57 +1978,192 @@ def initialize_triggers(self):
+
+        # All done!
+        content.write("</body></html>")
+        return content.getvalue()
+
+    def _write_div(self, content, markdown, class_name):
+        content.write('<div class="')
+        content.write(class_name)
+        content.write('">')
+        content.write(mdpopups.md2html(self.view, markdown))  # type: ignore
+        content.write("</div>\n")
 
     def handle_response(self, response, point):
-        if response is not None:
-            config = config_for_scope(self.view)
-            signatures = response.get("signatures")
-            activeSignature = response.get("activeSignature")
-            debug("got signatures, active is", len(signatures), activeSignature)
-            if len(signatures) > 0 and config:
-                signature = signatures[activeSignature]
-                debug("active signature", signature)
-                formatted = []
-                formatted.append(
-                    "```{}\n{}\n```".format(config.languageId, signature.get('label')))
-                params = signature.get('parameters')
-                if params is None:  # for pyls TODO create issue?
-                    params = signature.get('params')
-                debug("params", params)
-                for parameter in params:
-                    paramDocs = parameter.get('documentation')
-                    if paramDocs:
-                        formatted.append("**{}**\n".format(parameter.get('label')))
-                        formatted.append("* *{}*\n".format(paramDocs))
-
-                formatted.append(signature.get('documentation'))
-
-                mdpopups.show_popup(
-                    self.view,
-                    preserve_whitespace("\n".join(formatted)),
-                    css=".mdpopups .lsp_signature { margin: 4px; } .mdpopups p { margin: 0.1rem; }",
-                    md=True,
-                    flags=sublime.HIDE_ON_MOUSE_MOVE_AWAY,
-                    location=point,
-                    wrapper_class="lsp_signature",
-                    max_width=800)
+
+        if not response:
+            return
+        config = config_for_scope(self.view)
+        if not config:
+            debug("no config found for view", self.view.id())
+            return
+
+        # Fetch all the relevant data.
+        self.signatures = response.get("signatures", None)
+        x, y = self.view.viewport_extent()
+        debug("max width:", x, "max height:", y)
+        if self.signatures:
+            self.active_signature = response.get("activeSignature", -1)  # Optional.
+            self.active_parameter = response.get("activeParameter", -1)  # Optional.
+            self.view.show_popup(self._build_popup_content(), 0, point, int(x), int(y),
+                                 self._on_navigate, self._on_hide)
+        else:
+            self.view.show_popup('<html><body id="lsp"><div class="error">There are no signatures!</div></body></html>',
```

Comment:
This can be "too clever" sometimes, for example when a trigger character is not actually inside of a func call this'll pop up. Maybe just remove it?